static-ips, layer2: Add IP family validation

RamLavi opened the pull request:
## 📑 Description
This PR is supporting proper static-ip requests on layer2 subnets with dual stack, by enforcing that the "ips" argument at the "v1.multus-cni.io/default-network" annotation follow the ip family from the layer2 network.
This can be summed in this bullets:

* single stack ipv4 -> it should have just one ipv4 address
* dual stack -> it should have a pair of addresses one ipv4 and the other ipv6
* single stack ipv6 -> it should have just one ipv6 address

*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://issues.redhat.com/browse/CORENET-6255

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [ ] My code requires tests
- [ ] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Pod static IP allocation now enforces IP-family compatibility (IPv4-only, IPv6-only, dual-stack) for preconfigured addresses on primary layer2 networks; mismatches abort allocation with a clear error.

* **Bug Fixes**
  * IP-family validation failures are surfaced as allocation errors so they are recorded and block invalid assignments.

* **Tests**
  * Expanded unit and e2e coverage for single-/dual-stack success and failure, rollback behavior, event-based error checks, and parameterized network CIDR scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -261,6 +261,56 @@ func validateStaticIPRequest(netInfo util.NetInfo, network *nadapi.NetworkSelect
+		} else {
+			ipv4Count++
+		}
+	}
```
Check first if there are more thatn 2 ip addresses, so we fail fast without looping.

kyrtapz reviewed (changes requested):
e2e failures look legit

RamLavi commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -418,6 +475,9 @@ func allocatePodAnnotationWithRollback(
 				klog.Warningf("Failed parsing IPRequest %+v for pod %s: %v", network.IPRequest, podDesc, err)
 				return
 			}
+			if err = validateIPFamilyMatchesNetwork(netInfo, tentative.IPs); err != nil {
```
good idea, move it there.

RamLavi commented:
> e2e failures look legit

ah, yes. the test fails are indeed legit, I will need to refactor the existing duplicate IPs tests to make them check it more accurately.

RamLavi commented:
[Change](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/c666a48de4eba8d96211beb71dab963e2130290f..76eb7c5c6ed3c642fc28884429f11ab064b00922): Addressed @kyrtapz review, refactored e2e tests that failed due to the current change

kyrtapz reviewed (approved):
LGTM
@qinqon are you good with the latest version?

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -258,6 +258,51 @@ func validateStaticIPRequest(netInfo util.NetInfo, network *nadapi.NetworkSelect
+
+	ipv4Mode, ipv6Mode := netInfo.IPMode()
+	if ipv4Mode != requestedIPv4 || ipv6Mode != requestedIPv6 {
+		return fmt.Errorf("layer2 network IP family mismatch: network supports IPv4=%t IPv6=%t, but requested types IPv4=%t IPv6=%t: %w",
```
```suggestion
		return fmt.Errorf("layer2 network IP family mismatch: network supports IPv4=%t IPv6=%t, but requested addresses IPv4=%t IPv6=%t: %w",
```

not 100% sure this reads better though
